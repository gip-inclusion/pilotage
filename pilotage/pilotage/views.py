from django.shortcuts import render

def home(request):
    return render(request, 'pilotage/home.html')

def stats(request):
    return render(request, 'pilotage/statistiques.html')

def accessibilite(request):
    return render(request, 'pilotage/accessibilite.html')

def politique_de_confidentialite(request):
    return render(request, 'pilotage/politique_de_confidentialite.html')

def inscription_lettre_information(request):
    return render(request, 'pilotage/inscription_lettre_information.html')

def mentions_legales(request):
    return render(request, 'pilotage/mentions_legales.html')

def tableaux_de_bord_prives(request):
    return render(request, 'pilotage/tableaux_de_bord_prives.html')
