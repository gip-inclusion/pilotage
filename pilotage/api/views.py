import copy

from django.conf import settings
from django.db.models.enums import TextChoices
from django.http import (
    JsonResponse,
)

from pilotage.itoutils import metabase
from pilotage.itoutils.departments import DEPARTMENTS


class DataSetName(TextChoices):
    DI_SERVICES = "di_services", "data·inclusion - Services"
    DI_STRUCTURES = "di_structures", "data·inclusion - Structures"


QUERIES = {
    DataSetName.DI_SERVICES: {
        "database": 2,
        "type": "query",
        "query": {
            "source-table": 2052,
            "fields": [
                ["field", 59639, {"base-type": "type/Text"}],  # ID
                ["field", 59644, {"base-type": "type/Text"}],  # Structure ID
                ["field", 59647, {"base-type": "type/Text"}],  # Source
                ["field", 59614, {"base-type": "type/Text"}],  # Nom
                ["field", 59643, {"base-type": "type/*"}],  # Thematiques
                ["field", 59627, {"base-type": "type/*"}],  # Frais
                ["field", 59636, {"base-type": "type/*"}],  # Profils
                ["field", 59649, {"base-type": "type/Text"}],  # Commune
                ["field", 59623, {"base-type": "type/Text"}],  # Code Postal
                ["field", 59654, {"base-type": "type/Text"}],  # Code Insee
                ["field", 59615, {"base-type": "type/Text"}],  # Adresse
                ["field", 59629, {"base-type": "type/Float"}],  # Longitude
                ["field", 59651, {"base-type": "type/Float"}],  # Latitude
                ["field", 59624, {"base-type": "type/*"}],  # Modes Accueil
            ],
            "filter": [
                "starts-with",
                ["field", 59654, {"base-type": "type/Text"}],
                "000",  # Default to no-matchs
                {"case-sensitive": False},
            ],
        },
    },
    DataSetName.DI_STRUCTURES: {
        "database": 2,
        "type": "query",
        "query": {
            "source-table": 2051,
            "fields": [
                ["field", 59591, {"base-type": "type/Text"}],  # ID
                ["field", 59600, {"base-type": "type/Text"}],  # Nom
                ["field", 59608, {"base-type": "type/Text"}],  # Commune
                ["field", 59606, {"base-type": "type/Text"}],  # Code Postal
                ["field", 59588, {"base-type": "type/Text"}],  # Code Insee
                ["field", 59605, {"base-type": "type/Text"}],  # Adresse
                ["field", 59594, {"base-type": "type/Float"}],  # Longitude
                ["field", 59599, {"base-type": "type/Float"}],  # Latitude
                ["field", 59590, {"base-type": "type/Text"}],  # Typologie
            ],
            "filter": [
                "starts-with",
                ["field", 59588, {"base-type": "type/Text"}],
                "000",  # Default to no-matchs
                {"case-sensitive": False},
            ],
        },
    },
}


def dataset(request, name):
    if name not in DataSetName:
        return JsonResponse({"error": "Dataset doesn't exists"}, status=404)

    # TODO: Use a query-builder like function to avoid a global and copying the query
    query = copy.deepcopy(QUERIES[DataSetName(name)])
    if department := request.GET.get("department"):
        if department not in DEPARTMENTS:
            return JsonResponse({"error": "Department doesn't exists"}, status=400)
        query["query"]["filter"][2] = department
    else:
        return JsonResponse({"error": "Department is a required query parameter"}, status=400)

    return JsonResponse(
        metabase.Client(settings.METABASE_BASE_URL).fetch_dataset_results(query),
        safe=False,  # Metabase return a list
    )
