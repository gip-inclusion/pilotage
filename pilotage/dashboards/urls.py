from django.contrib import admin
from django.urls import path
from dashboards import views

app_name = "dashboards"
urlpatterns = [
    path(
        "tableaux-de-bord/",
        views.tableaux_de_bord_publics,
        name="tableaux_de_bord_publics",
    ),
    path(
        "tableaux-de-bord/<slug:slug>/",
        views.tableau_de_bord_public,
        name="tableau_de_bord_public",
    ),
]
