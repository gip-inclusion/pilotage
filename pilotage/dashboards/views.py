from django.shortcuts import render
from django.db.models import Prefetch

from pilotage.dashboards.models import Category, Dashboard

def tableaux_de_bord_publics(request):
    categories = Category.objects.all().prefetch_related(
        Prefetch(
            "dashboard_set",
            to_attr="dashboards",
            queryset=Dashboard.objects.filter(active=True),
        ),
    )
    return render(
        request,
        "dashboards/tableaux_de_bord_publics.html",
        context={"categories": categories},
    )


def tableau_de_bord_public(request, slug):
    dashboard = Dashboard.objects.get(slug=slug)

    return render(
        request,
        "dashboards/tableau_de_bord_public.html",
        context={"dashboard": dashboard, "slug": slug},
    )
