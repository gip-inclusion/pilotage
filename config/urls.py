from django.conf import settings
from django.contrib import admin
from django.urls import include, path
from pilotage.pilotage import urls as pilotage_urls
from pilotage.dashboards import urls as dashboards_urls

urlpatterns = [
    path(settings.PATH_TO_ADMIN, admin.site.urls),
    path("", include(pilotage_urls)),
    path("", include(dashboards_urls)),
]
