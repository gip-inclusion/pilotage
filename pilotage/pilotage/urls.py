"""
URL configuration for pilotage project.

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/4.2/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.contrib import admin
from django.urls import include, path
from . import views

app_name = 'pilotage'
urlpatterns = [
    path('admin/', admin.site.urls),
    path('', views.home, name="home"),
	path('dashboards/', include('dashboards.urls')),
    path('stats/', views.stats, name="statistiques"),
    path('accessibilite/', views.accessibilite, name="accessibilite"),
    path('mentions-legales/', views.mentions_legales, name="mentions_legales"),
    path('politique-de-confidentialite/', views.politique_de_confidentialite, name="politique_de_confidentialite"),
    path('inscription-lettre-information/', views.inscription_lettre_information, name="inscription_lettre_information"),
    path('tableaux-de-bord-prives/', views.tableaux_de_bord_prives, name="tableaux_de_bord_prives"),
]
