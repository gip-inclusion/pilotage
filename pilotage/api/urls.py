from django.urls import path

from pilotage.api import views


app_name = "api"
urlpatterns = [
    path("dataset/<slug:name>", views.dataset, name="dataset"),
]
