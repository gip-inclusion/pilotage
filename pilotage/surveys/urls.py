from django.urls import path

from pilotage.surveys import views


app_name = "surveys"
urlpatterns = [
    path("<slug:survey_name>/", views.start, name="start"),
    path("<slug:survey_name>/<uuid:answer_uid>/<str:step>", views.tunnel, name="tunnel"),
]
