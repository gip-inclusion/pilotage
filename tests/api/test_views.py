import random

import pytest
from django.urls import reverse

from pilotage.api.views import DataSetName
from pilotage.itoutils import departments


@pytest.mark.parametrize("dataset_name", DataSetName)
def test_dataset(mocker, faker, client, dataset_name):
    expected_data = [faker.pydict(allowed_types=[str, int])]
    mocker.patch("pilotage.itoutils.metabase.Client.fetch_dataset_results", return_value=expected_data)

    response = client.get(
        reverse("api:dataset", kwargs={"name": dataset_name}),
        query_params={"department": random.choice(list(departments.DEPARTMENTS.keys()))},
    )
    assert response.json() == expected_data


def test_dataset_with_unknown_name(client, snapshot):
    response = client.get(reverse("api:dataset", kwargs={"name": "unknown"}))
    assert response.status_code == 404
    assert response.json() == snapshot()


@pytest.mark.parametrize("dataset_name", DataSetName)
def test_dataset_without_department(snapshot, client, dataset_name):
    response = client.get(reverse("api:dataset", kwargs={"name": dataset_name}))
    assert response.status_code == 400
    assert response.json() == snapshot()


@pytest.mark.parametrize("dataset_name", DataSetName)
def test_dataset_with_unknown_department(snapshot, client, dataset_name):
    response = client.get(reverse("api:dataset", kwargs={"name": dataset_name}), query_params={"department": "00"})
    assert response.status_code == 400
    assert response.json() == snapshot()
