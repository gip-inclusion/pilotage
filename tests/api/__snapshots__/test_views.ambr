# serializer version: 1
# name: test_dataset_with_unknown_department[di_services]
  dict({
    'error': "Department doesn't exists",
  })
# ---
# name: test_dataset_with_unknown_department[di_structures]
  dict({
    'error': "Department doesn't exists",
  })
# ---
# name: test_dataset_with_unknown_name
  dict({
    'error': "Dataset doesn't exists",
  })
# ---
# name: test_dataset_without_department[di_services]
  dict({
    'error': 'Department is a required query parameter',
  })
# ---
# name: test_dataset_without_department[di_structures]
  dict({
    'error': 'Department is a required query parameter',
  })
# ---
